test/jdk/javax/swing/JFileChooser/FileSystemView/SystemIconTest.java

```diff
@@ -0,0 +1,105 @@
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8182043
+ * @summary Access to Windows Large Icons
+ * sun.awt.shell.ShellFolder
+ * @run main SystemIconTest
+ */
+import javax.swing.Icon;
+import javax.swing.ImageIcon;
+import javax.swing.filechooser.FileSystemView;
+import java.awt.image.MultiResolutionImage;
+import java.io.File;
+
+public class SystemIconTest {
+    static final FileSystemView fsv = FileSystemView.getFileSystemView();
+
+    public static void main(String[] args) {
+        testSystemIcon();
+        negativeTests();
+    }
+
+    static void testSystemIcon() {
+        String os = System.getProperty("os.name");
+        if (os.startsWith("Windows")) {
+            String windir = System.getenv("windir");
+            testSystemIcon(new File(windir), true);
+            testSystemIcon(new File(windir + "/explorer.exe"),
+                    true);
+        } else {
+            String homedir = System.getProperty("user.home");
+            testSystemIcon(new File(homedir), false);
+        }
+    }
+
+    static void negativeTests() {
+        Icon icon;
+        try {
+            icon = fsv.getSystemIcon(new File("."), -1, 16);
+            throw new RuntimeException("Negative size icon should throw invalid argument exception");
+        } catch (IllegalArgumentException iae) {
+            // Expected
+        }
+
+        icon = fsv.getSystemIcon(new File("thereisdefinitelynosuchfile.why"),
+                16, 16);
+        if (icon != null) {
+            throw new RuntimeException("Icons for files with invalid names should be null");
+        }
+    }
+
+    static void testSystemIcon(File file, boolean implComplete) {
+        int[] sizes = new int[] {16, 32, 48, 64, 128};
+        for (int size : sizes) {
+            ImageIcon icon = (ImageIcon) fsv.getSystemIcon(file, size, size);
+
+            //Enable below to see the icon
+            //JLabel label = new JLabel(icon);
+            //JOptionPane.showMessageDialog(null, label);
+
+            if (icon == null) {
+                throw new RuntimeException("icon is null!!!");
+            }
+
+            if (implComplete && icon.getIconWidth() != size) {
```

Comment:
Why the size of the returned images might be different? The spec state the size parameter in the getSystemIcon is the size of the icon in the userspace.

Reply:
The spec says that we are taking into consideration requested sizes but depending on the platform implementation exact match can not be guaranteed. On Windows it can so i'm checking for that.

Reply:
Is this requirement is so important? can we return an MRI(same as on Windows) which will have just one resolution? Otherwise what the user should do if the requested size was 32x32 but returned image will be 21x21? Paint the small icon or rescale it by the application?

Reply:
> Is this requirement is so important? can we return an MRI(same as on Windows) which will have just one resolution?

We might - when the implementation will be done on other platforms. Probably it will be done by me, probably - by someone else. Right now we return whatever we have so on Linux it is UIManager default icons for file or a folder (which is exactly what any file manager on Linux shows for any file and this is exactly what we promised in the method description). In the future it can change but for now it is all we can guarantee.

Reply:
This is my point I think we should update this implementation to always return MRI of the requested size, otherwise, the code example of this will look like this:
```
    Icon icon = fsv.getSystemIcon(file, width, height);
    if (icon.getIconWidth() != width && icon.getIconHeight() != height) {
        return scaleTheIconInTheSameWayAsBeforeTheFix(icon, width, height);
    } else if (icon instanceof ImageIcon) {
        ImageIcon imageIcon = (ImageIcon) icon;
        if (icon.getImage() instanceof MultiResolutionImage) {
            MultiResolutionImage mri = (MultiResolutionImage) icon.getImage();
            return  mri.getResolutionVariant(width, height);
        } else {
            return imageIcon;
        }
    } else {
        return icon;
    }
```
I pretty sure we can do better than the code above.